src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs

```diff
@@ -0,0 +1,505 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Security.Utilities;
+
+#if NET8_0_OR_GREATER
+using System.Buffers;
+using StringInput = System.ReadOnlySpan<char>;
+#else
+using StringInput = string;
+#endif
+
+/// <summary>
+/// A detection from <see cref="HighPerformanceScanner"/>.
+/// </summary>
+internal readonly record struct HighPerformanceDetection
+{
+    /// <summary>
+    /// The signature that identifies the pattern detected.
+    /// </summary>
+    public string Signature { get; }
+
+    /// <summary>
+    /// The start index in the input where the pattern was detected.
+    /// </summary>
+    public int Start { get; }
+
+    /// <summary>
+    /// The length of the  in the input that matched in the
+    /// </summary>
+    public int Length { get; }
+
+    public HighPerformanceDetection(string signature, int start, int length)
+    {
+        Signature = signature;
+        Start = start;
+        Length = length;
+    }
+}
+
+/// <summary>
+/// A pattern that <see cref="HighPerformanceScanner"/> can scan for. These are
+/// typically the precomputed set derived from the patterns that implement <see
+/// cref="IHighPerformanceScannableKey"/>.
+/// </summary>
+internal sealed partial class CompiledHighPerformancePattern
+{
+    /// <summary>
+    /// The signature that identifies this pattern. This is searched for as fast
+    /// as possible in the input before applying the regex.
+    /// </summary>
+    public string Signature { get; }
+
+    /// <summary>
+    /// The fixed number of characters that appear before the signature in a match.
+    /// </summary>
+    public int PreSignatureLength { get; }
+
+    /// <summary>
+    /// The minimum number of characters that can appear in a match.
+    /// </summary>
+    public int MinLength { get; }
+
+    /// <summary>
+    /// The maximum number of characters that can appear in a match.
+    /// </summary>
+    public int MaxLength { get; }
+
+    /// <summary>
+    /// The regular expression that matches the pattern near a signature.
```

Comment:
>near a signature [](http://example.com/codeflow?start=56&length=16)

Hm, strictly speaking our current rule set for the most precise patterns incorporate a signature. But in some circumstances (like the `sig=` case), the signature is elided at match refinement. So ok.

The pedant in me says use this:
`matches the pattern near or which incorporates a signature`

Reply:
That feels awkard to me. I don't think this distinction matters given here all the copious explanation of how this works everywhere near here.